diff --git a/hotspot/src/os/linux/vm/os_linux.cpp b/hotspot/src/os/linux/vm/os_linux.cpp
index 2d3880b363..77eb2ac3fd 100644
--- a/hotspot/src/os/linux/vm/os_linux.cpp
+++ b/hotspot/src/os/linux/vm/os_linux.cpp
@@ -93,11 +95,17 @@
 # include <semaphore.h>
 # include <fcntl.h>
 # include <string.h>
-# include <syscall.h>
 # include <sys/sysinfo.h>
+#ifndef __ANDROID__
 # include <gnu/libc-version.h>
+#endif
 # include <sys/ipc.h>
+#if !defined(__ANDROID__)
+# include <syscall.h>
 # include <sys/shm.h>
+#else
+# include <sys/syscall.h>
+#endif
 # include <link.h>
 # include <stdint.h>
 # include <inttypes.h>
@@ -2977,7 +3048,11 @@ extern "C" JNIEXPORT int fork1() { return fork(); }
 // Handle request to load libnuma symbol version 1.1 (API v1). If it fails
 // load symbol from base version instead.
 void* os::Linux::libnuma_dlsym(void* handle, const char *name) {
+#ifndef __ANDROID__
   void *f = dlvsym(handle, name, "libnuma_1.1");
+#else
+  void *f = NULL;
+#endif
   if (f == NULL) {
     f = dlsym(handle, name);
   }
@@ -5867,7 +5978,11 @@ bool os::is_thread_cpu_time_supported() {
 // Linux doesn't yet have a (official) notion of processor sets,
 // so just return the system wide load average.
 int os::loadavg(double loadavg[], int nelem) {
+#ifdef __ANDROID__
+  return -1;
+#else
   return ::getloadavg(loadavg, nelem);
+#endif // !__ANDROID__
 }
 
 void os::pause() {
diff --git a/common/autoconf/build-aux/config.sub b/common/autoconf/build-aux/config.sub
index cc958da946..162e27c064 100644
--- a/common/autoconf/build-aux/config.sub
+++ b/common/autoconf/build-aux/config.sub
@@ -30,7 +30,8 @@
 DIR=`dirname $0`
 
 # First, filter out everything that doesn't begin with "aarch64-"
-if ! echo $* | grep '^aarch64-' >/dev/null ; then
+# or the mobile platforms (ios and android)
+if ! echo $* | egrep "^aarch64-|-android" >/dev/null ; then
     . $DIR/autoconf-config.sub "$@"
     # autoconf-config.sub exits, so we never reach here, but just in
     # case we do:
@@ -41,6 +42,10 @@ while test $# -gt 0 ; do
     case $1 in 
         -- )   # Stop option processing
             shift; break ;;
+        *-android* )
+            echo $1
+            exit
+            ;;
         aarch64-* )
             config=`echo $1 | sed 's/^aarch64-/arm-/'`
             sub_args="$sub_args $config"
diff --git a/hotspot/src/cpu/aarch64/vm/immediate_aarch64.cpp b/hotspot/src/cpu/aarch64/vm/immediate_aarch64.cpp
index 9f606ff3..7703b0f7 100644
--- a/hotspot/src/cpu/aarch64/vm/immediate_aarch64.cpp
+++ b/hotspot/src/cpu/aarch64/vm/immediate_aarch64.cpp
@@ -122,8 +122,16 @@ static inline u_int32_t uimm(u_int32_t val, int hi, int lo)
 
 u_int64_t replicate(u_int64_t bits, int nbits, int count)
 {
+  assert(count > 0, "must be");
+  assert(nbits > 0, "must be");
+  assert(count * nbits <= 64, "must be");
+  // Special case nbits == 64 since the shift below with that nbits value
+  // would result in undefined behavior.
+  if (nbits == 64) {
+    return bits;
+  }
+
   u_int64_t result = 0;
-  // nbits may be 64 in which case we want mask to be -1
   u_int64_t mask = ones(nbits);
   for (int i = 0; i < count ; i++) {
     result <<= nbits;
